Feature: Flexible Y-Min and Y-Max settings.

thoj opened the pull request:
Possible Issues: #979 #4156 #3935 #2759

Y-Min and Y-Max is now string.
New usage for both Y-Min and Y-Max (Where X is a real number):

```
  >X Y-Max/Y-Max is auto if data is above X else X
  <X Y-Max/Y-Min is auto if data is below X else X
  =X Y-Max/Y-Min is scaled to current value +/- X
  ~X Y-Max/Y-Min is caled to average value +/- X
```

Example: Y-Min: <100 Y-Max: >200
If all points are within 100 and 200: Y-Min is 100 and Y-Max is 200
If some points are above 200: Y-Min is 100 and Y-Max is auto
If some points are below 100: Y-Min is auto and Y-Max is 200
if some points are below 100 and above 200: Y-Min and Y-Max is auto

Unit tests for new settings added

More Examples: 
In the following images you see temperature gauges that when using default auto-scaling is either noisy or show very little information because of a few values that are close to zero. 

I can't set Y-Min and Y-Max because the temperatures will wander to much over a few days and would fall outside of Y-Min and Y-Max pretty quickly. 

![autoscale-example1](https://cloud.githubusercontent.com/assets/153440/17615886/75e3d0ee-6071-11e6-9a3d-d3764550c90c.PNG)
![autoscale-example2](https://cloud.githubusercontent.com/assets/153440/17615887/7a3a43da-6071-11e6-84d0-bd5067f9cb3f.PNG)
![autoscale-example3](https://cloud.githubusercontent.com/assets/153440/17615893/7e8436d0-6071-11e6-8c88-ffa6f853538f.PNG)
![autoscale-example4](https://cloud.githubusercontent.com/assets/153440/17615897/8398e36e-6071-11e6-83a5-b14ea5d64168.PNG)
![autoscale-display](https://cloud.githubusercontent.com/assets/153440/17615905/8853b6fe-6071-11e6-98d2-28dd49c1fb62.PNG)

lpalm commented:
This is great.

The "<X" and ">X" functions could potentially be applied directly to the Y-Min and Y-Max, respectively. This would allow for fine control over the min and max ranges separately. It could be useful when certain fixed Y values are important (e.g. 0 or some threshold), but still allowing points that cross those Y values to be displayed.

Example:
Having a Level 1 threshold of 600 ("minimum") degrees and a level 2 threshold of 800 ("maximum") that are always visible on the graph, but still allowing points <600 or >800 to be visible if they're present. Having a minimum span of 200 would almost always work, but would hide the thresholds if we zoomed into a range where all points are in the range 1000-1100, for example.

lpalm commented:
Yeah, I figured we might as well allow strings in the Y-Min/Max fields to save space.

thoj commented:
I'll se what I can do :) Thanks for your ideas :)

thoj committed:
Feature: Flexible Y-Min and Y-Max settings.

Y-Min and Y-Max is now string.
New usage for both Y-Min and Y-Max (Where X is a real number):
  >X Y-Max/Y-Max is auto if data is above X else X
  <X Y-Max/Y-Min is auto if data is below X else X
  =X Y-Max/Y-Min is scaled to current value +/- X
  ~X Y-Max/Y-Min is caled to average value +/- X

Example: Y-Min: <100 Y-Max: >200
If all points are within 100 and 200: Y-Min is 100 and Y-Max is 200
If some points are above 200: Y-Min is 100 and Y-Max is auto
If some points are below 100: Y-Min is auto and Y-Max is 200
if some points are below 100 and above 200: Y-Min and Y-Max is auto

Tests for new settings added

thoj commented:
@lpalm @torkelo Did som refactoring and added unit tests. Now just using Y-Min and Y-Max.

 @lpalm if you look at the first example image i think this is what you wanted?

p.s. This is the first time I've done unit tests in angular and it took me some time to gork, hope it's done right.

lpalm commented on public/app/plugins/panel/graph/graph.js:
```diff
@@ -352,6 +352,45 @@ function (angular, $, moment, _, kbn, GraphTooltip) {
+        function autoscaleSpanOverride(yaxis, data, options) {
+          var m, op, num, precision;
+          if (yaxis.min != null && data != null) {
+            m = yaxis.min.match(/([<=>~]*)\W*(\d+(\.\d+)?)/);
```
Should we do:
/([<=>~]?)\W*(\d+(\.\d+)?)/

Instead of:
/([<=>~]*)\W*(\d+(\.\d+)?)/

thoj replied:
@lpalm I have a bit of a difficult time figuring out where <= and >= would be useful? Could you explain one case for me? 

I agree that the regex could be a lot better. I think we need \W* because \W would enforce a white space.

lpalm replied:
It wouldn't be useful. Rather, it may be nice to treat >= as > and <= as < instead of just not working. Don't feel strongly about it.

Apologies about the \W*, looks like github treated the asterisk as an indication to italicize my text! I've edited it to be in block quotes now.

lpalm commented:
I'm not super familiar with the Grafana code base and style but LGTM! Thanks for putting this in, @thoj. And yes, that screenshot is exactly what I had in mind.

thoj committed:
Use parseFloat instead of implicit conversion

torkelo commented:
sorry, this PR was refactored a bit in this PR (https://github.com/grafana/grafana/pull/6051) , I hit squash merge commit without thinking that this PR would not then be automatically detected as merged  by github. 

Anyway, the feature is in master now, thank you for making Grafana better :heart: !

torkelo commented:
closing this as this PR has drifted a bit too far from master. 

Need to figure out a UI for the expressions so they are discoverable (and not secret things you need to know about)
